### ISSUE TITLE
Calling perform_action is not the same as hitting the key (potentially search only scope)

### ISSUE THREAD

**Issue by mitchcapper** [wez/wezterm#3102]
### What Operating System(s) are you seeing this problem on?

Windows

### WezTerm version

20230212-103106-a5c2b1f3

### Did you try the latest nightly build to see if the issue is better (or worse!) than your current version?

Yes, and I updated the version box above to show the version of the nightly that I tried

### Describe the bug

After my second example was shown to be invalid, it is possible this is only  search related.  I have not yet dug into the code to try and follow the event action in more detail.
One would expect a call to perform_action to be pretty similar to actually hitting a bound key press to the same action but it seems there are subtle differences.  

This is likely what is biting a solution like: https://github.com/wez/wezterm/issues/1988 from working.

### To Reproduce

I took it a step further trying a few different things binding a key to:
```lua
wezterm.action_callback(function(win, pane)
win:perform_action(wezterm.action.CopyMode("ClearPattern"), pane)
action = wezterm.action_callback(function(win, pane)
win:perform_action(act.Search({CaseInSensitiveString = ""}), pane)
wezterm.sleep_ms(4300)
win:perform_action(wezterm.action.CopyMode("ClearPattern"), pane)
win:perform_action(wezterm.action.CopyMode("ClearPattern"), win:active_pane())
wezterm.log_info("DONE SEARCH2 active table: ", win:active_key_table())
end)
```

I figured maybe the pane changing was the issue, tried with the delay to make sure search was fully loaded, etc but none of it clears the search box.  Hitting control + u does though, even if its bound to the exact same ClearPattern call.

This is not just ClearPattern but also things like `win:perform_action(wezterm.action.CopyMode("PriorMatch"), pane)`

~~Another example I found was using @jakehl 's  https://github.com/twilsoft/wezmode  it uses just key tables to change its actions and display.    I found that if I activate a key table,  then active search it switches to the search_mode key table and the wezmode binds do not work.   I did find if I hit the key combination to activate the wezmode key table again even though the active_key_table stays on `search_mode` the key presses now register properly.    However if I use perform action with a sleep after the search runs like
`win:perform_action(wezterm.action.ActivateKeyTable({name = "WezmodeTable"}), pane)`  it doesn't work even though would in theory be the same.~~

### Configuration

see above

### Expected Behavior

perform_action to do the same thing as a key press bound to that action

### Logs

logs only have the console messages I was printing, although not sure if one should turn the log level up first for this.

**Comment by jakehl** [comment]
I'm not sure if there is a bug here. I think what you're experiencing is that you are calling `ActivateKeyTable` with different parameters depending on weather it's via a keypress or not. (see [here](https://wezfurlong.org/wezterm/config/lua/keyassignment/ActivateKeyTable.html) for the docs, particularly pay attention to `until_unknown`, `one_shot`, `prevent_fallback` and `replace_current`)

I'll try to explain below, please feel free to correct me if I'm wrong.

>  I found that if I activate a key table, then active search it switches to the search_mode key table and the wezmode binds do not work. I did find if I hit the key combination to activate the wezmode key table again even though the active_key_table stays on search_mode the key presses now register properly.

My understanding is that this is intended behavior. One key table can be active at any one time and they 'stack'. So as you pop a key table off the stack it should return to the previous table. Add to this, that by default, the `prevent_fallback` value on `ActivateKeyTable` is `false` which means that when your search mode table is active and wezmode hasn't been dismissed, it will correctly engage wezmode bindings if there aren't any matches on the current table.

So unless you have a custom keybind to activate the search table, it's likely that the default is calling `ActivateKeyTable` with parameters different to what you are calling it with programmatically. the `one_shot` setting particularly might be one to pay attention to here, as it's enabled by default, however I'd imagine the default binding probably sets this to false as this mode seems to stick until exited.

**Comment by mitchcapper** [comment]
Thanks @jakehl you are 100% right. The second example using your plugin the one_shot was key.  I actually noticed the intermittent success during testing of the first press (not realizing that was the factor at the time as was using reverse search so would often do forward first) and noted it to investigate later.   When I noticed the ClearPattern failures I thought this was a second example of the same issue.

I do believe the first example here does still show an issue unrelated to the activekeytable.   It uses `act.Search` to start search mode (which works) sleeps for a period of time then calls both ```win:perform_action(wezterm.action.CopyMode("ClearPattern"), win:active_pane())
win:perform_action(wezterm.action.CopyMode("ClearPattern"), pane)``` without error.   Neither of which have any effect.  Manually using a key bind during that sleep period bound to the same action `action = wezterm.action.CopyMode("ClearPattern")`  does clear the search box.
